src/coreclr/System.Private.CoreLib/src/System/Array.CoreCLR.cs

```diff
@@ -381,8 +381,43 @@ private unsafe bool IsValueOfElementType(object value)
         // if this is an array of value classes and that value class has a default constructor
         // then this calls this default constructor on every element in the value class array.
         // otherwise this is a no-op.  Generally this method is called automatically by the compiler
+        public unsafe void Initialize()
+        {
+            MethodTable* pArrayMT = RuntimeHelpers.GetMethodTable(this);
+            var thElem = pArrayMT->GetArrayElementTypeHandle();
+            if (thElem.IsTypeDesc)
+            {
+                return;
+            }
+
+            var pElemMT = thElem.AsMethodTable();
+            if (!pElemMT->HasDefaultConstructor || !pElemMT->IsValueType)
+            {
+                return;
+            }
+
+            ref byte arrayRef = ref MemoryMarshal.GetArrayDataReference(this);
+            delegate*<ref byte, MethodTable*, void> ctorFtn = GetConstructorSlot(pElemMT);
+
+            nuint offset = 0;
+            for (var i = 0; i < Length; i++)
+            {
+                // Since GetConstructorSlot() is not idempotent and may have returned
+                // a non-optimal entry-point the first time round.
+                if (i == 1)
+                {
+                    ctorFtn = GetConstructorSlot(pElemMT);
+                }
+
+                ctorFtn(ref Unsafe.Add(ref arrayRef, offset), pElemMT);
+                offset += pArrayMT->ComponentSize;
+            }
+
+            GC.KeepAlive(this);
```

Comment:
This is not necessary. arrayRef is going to keeping the array alive.

Reply:
@jkotas in a hypothetical case where e.g. method is inlined and JIT performed some new loop-related optimization so it don't perform `Length` check for the last iteration (e.g. unrolled) - isn't `offset += pArrayMT->ComponentSize;` potentially dangerous? `offset` is not a field.

Reply:
I agree that it is potentially dangerous as written. It would be better to keep the element size in a local like what is in the code snippet from my suggestion (avoids the deterrence and widening in each iteration) that should eliminate the need for the KeepAlive here.